Adding angle utilities

cvuosalo opened the pull request:
#### PR description:

In many spots in CMSSW, code like the following appears:
```
if (pointphi >= twopi)
     pointphi -= twopi;
if(pointphi < 0)
     pointphi += twopi;
```
The intent is to constrain the angle to the range 0<= angle < 2pi, but the code, as in this example, sometimes doesn't handle the case of an |angle| >= 4pi.

`DataFormats/Math/interface/deltaPhi.h` provides a `reduceRange()` function for this purpose, but tests show this function is not as fast as it could be, and it suffers from sensitivity in its `if` statement to small deviations from pi.

This PR adds a new function `make0to2pi`, that can be significantly faster than `reduceRange()` and that greatly reduces value drift due to floating-point comparisons. This new function is added without changing the implementation of `reduceRange()` because changing `reduceRange()` could cause slight differences in program behavior. Usage of `reduceRange()` can later be replaced on a case-by-case basis where better performance and precision is required.

The `Phi` class in `DataFormats/GeometryVector/interface/Phi.h` represents an angle that maintains a value from -pi to pi. This PR enhances this class in a backwardly compatible fashion to allow the option of keeping the angle in the range from 0 to 2pi.

In addition, the generic angle utilities in the `GeantUnits.h` file are moved to `deltaPhi.h` because they are not Geant-specific and should be usable without subjecting code to Geant conventions.

#### PR validation:

A unit test program is included in this PR, and it verifies basic functions of the code and gives some rough performance measures.

Some performance results comparing `reduceRange()` and `make0to2pi` are shown below.

The entry for the `Phi` templates in the `classes_def.xml` file uses ClassVersion 3 because the revised template is effectively a new class. No problems with the ROOT dictionary for `Phi` were observed. Using baseline CMSSW_10_6_0_pre2 code without this PR, I wrote a `Phi` variable to a ROOT file. Then, with the new PR code, I could successfully read the `Phi` variable out of that same ROOT file, so there appears to be backwards compatibility in reading old ROOT files containing the `Phi` class. However, a scan of CMSSW finds no cases where `Phi` variables were written to ROOT files, so it doesn't look like the ability to read the old version of `Phi` variables in ROOT files is even needed.

No backport is needed.

cvuosalo committed:
Adding utilities for constraining angles to range 0 to 2pi

cvuosalo committed:
Fix formatting

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26227/8851

- This PR adds an extra 20KB to repository

cmsbuild commented:
A new Pull Request was created by  @cvuosalo (Carl Vuosalo) for master.

It involves the following packages:

DataFormats/Math


@cmsbuild, @perrotta, @slava77 can you please review it and eventually sign? Thanks.
@makortel, @felicepantaleo, @rovere this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cvuosalo commented:
I want to suggest a faster but more complex implementation of the `make0to2pi` function, compared to the version in this PR:
```
template <class valType>
inline constexpr valType make0to2pi(valType ang)
{
  constexpr valType twoPi = 2._pi;
  // if statements arranged to promote faster performance
  if (ang < 0.) {
    if (ang >= -twoPi)
      ang += twoPi;
    else ang = fmod(ang, twoPi) + twoPi;
  }
  else if (ang >= twoPi) {
    ang = fmod(ang, twoPi);
  }
  return (ang);
}
```
The `if` statements allow avoiding the function call if it's not necessary, but this version is more complex and could potentially have a subtle bug.
Reviewers, please let me know if you think I should replace the simple implementation with this more complex one.

cvuosalo commented:
@ianna: Please review and let us know if you approve.
@bsunanda: FYI

cvuosalo commented:
@VinInn, @fwyzard: If you have time and would like to comment, we would appreciate your thoughts on the following questions. For angle values that need to be kept in the range [0, 2pi), what coding approach best balances reliability, developer convenience, and performance? Does this PR achieve this balance, or could it be improved?

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 533a9ee4efe7cf8757445f7e0e4ee36457743405
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26227/33694/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26227/33694/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3114829
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3114631
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

VinInn commented:
Sorry, there are already N version of it.
What's wrong with 
https://cmssdt.cern.ch/dxr/CMSSW/source/DataFormats/Math/interface/deltaPhi.h#14
?

VinInn commented:
A variant
https://cmssdt.cern.ch/dxr/CMSSW/source/DataFormats/Math/interface/normalizedPhi.h
(even more useful as moves one angle in a range around another)

cvuosalo commented:
I will do performance tests comparing different versions of the angle-constraining function and then post the results.

cvuosalo commented:
hold

cmsbuild commented:
Pull request has been put on hold by  @cvuosalo
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

cvuosalo committed:
Move generic angle utilities out of GeantUnits.h and into deltaPhi.h

cmsbuild commented:
The code-checks are being triggered in jenkins.

cvuosalo committed:
Move Angle0to2pi into deltaPhi.h

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26227/8979

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #26227 was updated. @cmsbuild, @perrotta, @slava77 can you please check and sign again.

cvuosalo commented:
Performance results comparing old `reduceRange()` and new `make0to2pi`. To make them equivalent, 2pi was added to the output of `reduceRange()` if it was negative.
The tests involved about a billion iterations of repeatedly adding plus or minus 1 degree to an angle and then calling the function to constrain it to 0-2pi range. The number of iterations is an exact multiple of 180 but not 360, so when the increment is 1 degree, the final angle value should be pi exactly. Times are wallclock times in seconds. Reference value of pi is 3.14159265358979323.
```
                        Time and Final Value
Function      1*, double               | -1* double 
reduceRange   7.2   3.141592495204069  |  7.0   3.141592808929663  
make0to2pi    3.8   3.141592653589782  |  3.7   3.141592653589829

                        Time and Final Value
Function      1*, long double           | -1* long double 
reduceRange   18.3   3.141592805717177  |  18.0   3.14159280396822
make0to2pi    14.1   3.141592653589793  |  14.4   3.141592653589793
```

`make0to2pi` is significantly faster and more precise than `reduceRange` for small increments. Performance results for a large increment like 7.77pi are more equal because `make0to2pi` cannot avoid calling the expensive `fmod` function.

cvuosalo commented:
unhold

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: ab1cfde29266059fc6c5ed9351d57ac4b44daa0d

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26227/33838/summary.html

I found follow errors while testing this PR

**Failed tests:  Build**

* **Build**:

I found compilation error when building: 

<pre>>> Package DetectorDescription/Core built
>> Entering Package DetectorDescription/Core
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-03-28-1100/src/DetectorDescription/Core/test/testRunner.cpp 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-03-28-1100/src/DetectorDescription/Core/test/DDUnits.cppunit.cc 
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-03-28-1100/src/DetectorDescription/Core/test/DDUnits.cppunit.cc: In member function 'void testDDUnits::checkUnits()':
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-03-28-1100/src/DetectorDescription/Core/test/DDUnits.cppunit.cc:31:31: error: 'piRadians' was not declared in this scope
   std::cout << "\nMy pi: " << piRadians
                               ^~~~~~~~~
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-03-28-1100/src/DetectorDescription/Core/test/DDUnits.cppunit.cc:31:31: note: suggested alternative:
In file included from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-03-28-1100/src/DataFormats/Math/interface/GeantUnits.h:10:0,
                 from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-03-28-1100/src/DetectorDescription/Core/test/DDUnits.cppunit.cc:3:
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

cvuosalo committed:
Add 'using namespace angle_units;' needed by files that include GeantUnits.h

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26227/8998

- This PR adds an extra 24KB to repository

- Found files with invalid states:
  - DataFormats/Math/interface/Angle0to2pi.h:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405
    - Deleted: ab1cfde29266059fc6c5ed9351d57ac4b44daa0d

cmsbuild commented:
Pull request #26227 was updated. @cmsbuild, @perrotta, @slava77 can you please check and sign again.

VinInn commented:
CMS coding conventions require that non trivial classes to be hosted in their own header file.
Please move ```Angle0to2pi``` in its own header file.

VinInn commented:
btw is ```Angle0to2pi```
not very similar to old
https://github.com/cms-sw/cmssw/blob/master/DataFormats/GeometryVector/interface/Phi.h
?

cvuosalo committed:
Split off Angle0to2pi into its own header file

cvuosalo commented:
@VinInn: Thanks for finding Phi. I will check to see where it is used. These two implementations will probably need to be merged.

cvuosalo committed:
Added default constructor

cvuosalo committed:
Fix Angle0to2pi unit test

cvuosalo committed:
Enhanced Phi class to support 0-2pi range. Sped up make0To2pi function

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26227/9311

- This PR adds an extra 40KB to repository

- Found files with invalid states:
  - DataFormats/Math/interface/Angle0to2pi.h:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae, d38e1fc565d8ba3e6ed5240b976f036d6391ce40
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 2fe825d5791331dbeebc90b73f8cc35ec8812d85
    - Deleted: ab1cfde29266059fc6c5ed9351d57ac4b44daa0d, 590c9dbe79197d5d9894f92b30446bfb427948de
  - DataFormats/Math/test/Angle0to2piTest.cc:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 69a8c3cfd6f7f36591992b07de90e028ee6972a8, 71953ee36ac3f5caeed01e74da398805f081604c, 2fe825d5791331dbeebc90b73f8cc35ec8812d85, 52c4f4402ab72d079af3b46c3edbd50a17127e8e
    - Deleted: 590c9dbe79197d5d9894f92b30446bfb427948de

cmsbuild commented:
Pull request #26227 was updated. @perrotta, @civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

cvuosalo committed:
Fix unnecessary change

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26227/9313

- This PR adds an extra 40KB to repository

- Found files with invalid states:
  - DataFormats/Math/interface/Angle0to2pi.h:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae, d38e1fc565d8ba3e6ed5240b976f036d6391ce40
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 2fe825d5791331dbeebc90b73f8cc35ec8812d85
    - Deleted: ab1cfde29266059fc6c5ed9351d57ac4b44daa0d, 590c9dbe79197d5d9894f92b30446bfb427948de
  - DataFormats/Math/test/Angle0to2piTest.cc:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 69a8c3cfd6f7f36591992b07de90e028ee6972a8, 71953ee36ac3f5caeed01e74da398805f081604c, 2fe825d5791331dbeebc90b73f8cc35ec8812d85, 52c4f4402ab72d079af3b46c3edbd50a17127e8e
    - Deleted: 590c9dbe79197d5d9894f92b30446bfb427948de

cmsbuild commented:
Pull request #26227 was updated. @perrotta, @civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

cvuosalo commented:
Updated Performance Results
Again for about a billion iterations of repeated increments, the new `make0To2pi()` function is usually faster than `reduceRange()` in ` DataFormats/Math/interface/deltaPhi.h`. The best and worst performance results for  `make0To2pi()` are:

```
Function            Time [s]   Type of test
-----------         --------   ----------------   
reduceRange         7.0        1 degree decrement with double arithmetic
make0To2pi          3.7
=====================================================================
reduceRange         13.8       7.77pi decrement with double arithmetic
make0To2pi          14.7
```

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: c1b271df76a6b914fad0e9efeef1233fa995e258
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26227/34236/summary.html
Additional comment: Compilation Warnings: Yes

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26227/34236/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 13382 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3142783
 - DQMHistoTests: Total failures: 29533
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3113053
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.317 KiB( 31 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): 0.317 KiB        SiStrip/MechanicalView
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

cvuosalo commented:
Using a more accurate value of pi may have caused comparison differences. I will restore use of the old value of pi to try to eliminate the differences.
I will also remove `using namespace` from header files where it's possible.

cvuosalo committed:
Remove 'using namespace' from header files. Restore use of old pi values in -pi to pi version of Phi::normalize function.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26227/9357

- This PR adds an extra 76KB to repository

- Found files with invalid states:
  - DataFormats/Math/interface/Angle0to2pi.h:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae, d38e1fc565d8ba3e6ed5240b976f036d6391ce40
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 2fe825d5791331dbeebc90b73f8cc35ec8812d85
    - Deleted: ab1cfde29266059fc6c5ed9351d57ac4b44daa0d, 590c9dbe79197d5d9894f92b30446bfb427948de
  - DataFormats/Math/test/Angle0to2piTest.cc:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 69a8c3cfd6f7f36591992b07de90e028ee6972a8, 71953ee36ac3f5caeed01e74da398805f081604c, 2fe825d5791331dbeebc90b73f8cc35ec8812d85, 52c4f4402ab72d079af3b46c3edbd50a17127e8e
    - Deleted: 590c9dbe79197d5d9894f92b30446bfb427948de

cmsbuild commented:
Pull request #26227 was updated. @perrotta, @civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 06f6e771e8bdecb8bf5ce9c93dc7a91b380b2455

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26227/34286/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  HeaderConsistency**

* **Build**:

I found compilation error when building: 

<pre>>> Building shared library tmp/slc7_amd64_gcc700/src/AnalysisDataFormats/Egamma/src/AnalysisDataFormatsEgamma/libAnalysisDataFormatsEgamma.so
Copying tmp/slc7_amd64_gcc700/src/AnalysisDataFormats/Egamma/src/AnalysisDataFormatsEgamma/libAnalysisDataFormatsEgamma.so to productstore area:
>> Checking EDM Class Version for src/AnalysisDataFormats/Egamma/src/classes_def.xml in libAnalysisDataFormatsEgamma.so
@@@@ ----> OK  EDM Class Version 
>> Checking EDM Class Transients in libAnalysisDataFormatsEgamma.so
DataFormatsGeometryVector_xr dictionary forward declarations' payload:5:16: error: too few template parameters in template redeclaration
namespace Geom{template <class T> class __attribute__((annotate("$clingAutoload$DataFormats/GeometryVector/interface/Phi.h")))  __attribute__((annotate("$clingAutoload$DataFormats/GeometryVector/interface/Basic3DVector.h")))  Phi;
               ^~~~~~~~~~~~~~~~~~
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-04-19-1100/src/DataFormats/GeometryVector/interface/Phi.h:49:3: note: previous template declaration is here
  template <class T1, PhiRange range = PhiRange::MinusPiToPi>
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

cvuosalo committed:
Move 'using piRadians' to correct namespace for backward compatibility

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26227/9359

- This PR adds an extra 76KB to repository

- Found files with invalid states:
  - DataFormats/Math/interface/Angle0to2pi.h:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae, d38e1fc565d8ba3e6ed5240b976f036d6391ce40
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 2fe825d5791331dbeebc90b73f8cc35ec8812d85
    - Deleted: ab1cfde29266059fc6c5ed9351d57ac4b44daa0d, 590c9dbe79197d5d9894f92b30446bfb427948de
  - DataFormats/Math/test/Angle0to2piTest.cc:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 69a8c3cfd6f7f36591992b07de90e028ee6972a8, 71953ee36ac3f5caeed01e74da398805f081604c, 2fe825d5791331dbeebc90b73f8cc35ec8812d85, 52c4f4402ab72d079af3b46c3edbd50a17127e8e
    - Deleted: 590c9dbe79197d5d9894f92b30446bfb427948de

cmsbuild commented:
Pull request #26227 was updated. @perrotta, @civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 13f74feb2773751ad75c0d1c3d318b2edff22356
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26227/34289/summary.html
Additional comment: Compilation Warnings: Yes

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26227/34289/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 5746 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3211964
 - DQMHistoTests: Total failures: 7776
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3203984
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

cvuosalo commented:
Clang generates a warning for the `Phi<>` class, but this warning does not seem to be an actual problem:
```
>> Building LCG reflex dict from header file src/DataFormats/GeometryVector/src/classes.h
  Warning: Forward declarations of templates with enums as template parameters. The responsible class is: NonTypeTemplateParmDecl 0x2b037b8 </build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-04-19-2300/src/DataFormats/GeometryVector/interface/Phi.h:49:23, col:50> col:32 referenced 'enum Geom::PhiRange' depth 0 index 1 range
 `-TemplateArgument expr
  `-DeclRefExpr 0x2b03770 <col:40, col:50> 'enum Geom::PhiRange' EnumConstant 0x2b01180 'MinusPiToPi' 'enum Geom::PhiRange'

  Warning: Problems with arguments for forward declaration of class Phi
   Warning: Forward declarations of templates with enums as template parameters. The responsible class is: NonTypeTemplateParmDecl 0x2b037b8 </build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-04-19-2300/src/DataFormats/GeometryVector/interface/Phi.h:49:23, col:50> col:32 referenced 'enum Geom::PhiRange' depth 0 index 1 range
 `-TemplateArgument expr
  `-DeclRefExpr 0x2b03770 <col:40, col:50> 'enum Geom::PhiRange' EnumConstant 0x2b01180 'MinusPiToPi' 'enum Geom::PhiRange'

  Warning: Problems with arguments for forward declaration of class Phi
```

Following instructions from Philippe Canal, I was able to instantiate `Phi` variables in interactive ROOT, which Philippe said shows the class works OK, so the warning can be disregarded.

cvuosalo commented:
The comparison test differences are almost entirely in two upgrade simulation workflows where events are re-generated. With low statistics, the comparisons show minor statistical fluctuations due to random differences in the generated events. I don't think that this PR causes any significant differences.

cvuosalo commented:
+1

cvuosalo commented:
@perrotta, @civanch, @kpedro88, @slava77: I have addressed all issues with this PR that I'm aware of. Could you please review it and let me know if I missed anything? Thanks.

When this PR is merged, I will be recommending replacing CLHEP::deg angle units with the _deg literals and functions from `DataFormats/Math/interface/deltaPhi.h`. Also the enhanced `Phi` class can be used for any angles that must be kept within the ranges of -pi to pi or 0 to 2pi.

kpedro88 commented on DataFormats/GeometryVector/test/PhiTest.cc:
```diff
@@ -0,0 +1,230 @@
+
+#include <iostream>
+#include <iomanip>
+#include <math.h>
```
use `cmath` rather than `math.h`

cvuosalo replied:
OK. I will wait for Reco and Sim to review before making this little change to the test program, in case they want more changes.

cvuosalo replied:
cmath include fixed now.

kpedro88 commented:
@cvuosalo regarding comparison differences: is your suggestion that this PR causes some change in the random number sequence in the SIM step? It seems odd to me that only a few workflows are affected...

cvuosalo commented:
@kpedro88: I don't think this PR itself caused any change. I performed a test to understand the differences. I tested two copies of CMSSW_10_6_0_pre2. In one I did `git cms-addpkg` of the packages changed by this PR, but without the PR changes, while I did nothing more than `scram b` in the other. I ran workflow `21234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D21` in each and did the comparison. There were differences in the comparison plots like those seen above, even though there was no change in code. My guess is that the SIM random number sequence changed between the two runs of that workflow.

kpedro88 commented:
@cvuosalo I have seen this kind of thing happen before. It may have to do changes in the order of the machine instructions generated by the compiler, which can change the order of random number calls. This is very difficult to debug. I do find it a bit odd that primarily Phase 2 workflows 20034.0 and 21234.0 are affected - but I think these are the only matrix workflows with the v8 HGCal, so maybe it specifically affects the SD code for that version (which is different from the SD code used in newer versions).

perrotta commented on Geometry/HGCalGeometry/src/FastTimeGeometry.cc:
```diff
@@ -125,7 +125,7 @@ DetId FastTimeGeometry::getClosestCell(const GlobalPoint& r) const {
     etaZPhi = topology().dddConstants().getZPhi(zz,r.phi());
   } else {
-    double phi = (zside > 0) ? r.phi() : atan2(r.y(),-r.x());
+    double phi = (zside > 0) ? static_cast<double>(r.phi()) : atan2(r.y(),-r.x());
```
What is the purpose of explicitly adding this ```static_cast``` (and only here, not in L126, e.g.)?

cvuosalo replied:
The conditional operator `?` is resolved at compile time. It requires its two results to be the same type. Because the `Phi` class has various conversions and `atan2` is a template function, there is a compile-time ambiguity in the desired type of the results, though it is obvious to the reader. The static cast eliminates the ambiguity for the compiler.

This problem only occurs with the conditional operator, and probably only when the two results of the operator are classes or templates that can take on various types. If this statement were written as an if-then statement, there would be no problem. I checked out all the dependencies that use the `Phi` class and compiled all that code, and this line of code was the only one where there was any type ambiguity.

Everywhere else in the code, variables of the `Phi` class are automatically converted into the correct type at execute time, as far as I can tell.

I will add a comment here to explain the cast.

perrotta commented on DataFormats/GeometryVector/src/classes_def.xml:
```diff
@@ -14,15 +14,15 @@
   <class name="Geom::Theta<float>" ClassVersion="10">
    <version ClassVersion="10" checksum="2498739044"/>
   </class>
-  <class name="Geom::Phi<double>" ClassVersion="10">
```
Not clear to me what happens to these ```Geom::Phi<T>```, which are still used in, e.g., ```Basic3DVector.h```

cvuosalo replied:
@perrotta:
`Geom::Phi<T>` is now equivalent to `Geom::Phi<T, Geom::PhiRange::MinusPiToPi>` since `MinusPiToPi` is the default value for the second parameter.
I tested the dictionary in ROOT, and it recognizes `Geom::Phi<float>` and `Geom::Phi<double>`.
Do you think there will be any problems?

perrotta replied:
Ok, thank you for pointing out the default value for the second parameter!

cvuosalo replied:
I intended the enhanced `Phi` class to be fully backwardly compatible, so old code would be behave as before. The only exception was the static cast I had to add, as explained above.

slava77 replied:
Is the old Geom::Phi persisted anywhere?
If it is, there should be an effort made to read the old data in a compatible way.

cvuosalo replied:
The old version of `Geom::Phi` does not appear to be persisted, and even if it was, it can be read by the new code.

cvuosalo committed:
Comment and math include fixes

cmsbuild commented:
The code-checks are being triggered in jenkins.

cvuosalo commented on DataFormats/GeometryVector/interface/Phi.h:
```diff
@@ -11,125 +12,167 @@ namespace Geom {
+    // There may be cases of ambiguities because of multiple possible
+    // conversions, in which case explicit casts must be used.
+    // The constructor provides range checking and normalization,
+    // e.g. the value of Pi(2*pi()+1) is 1
```
Thanks for finding this. It is fixed now.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26227/9505

- This PR adds an extra 76KB to repository

- Found files with invalid states:
  - DataFormats/Math/interface/Angle0to2pi.h:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae, d38e1fc565d8ba3e6ed5240b976f036d6391ce40
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 2fe825d5791331dbeebc90b73f8cc35ec8812d85
    - Deleted: ab1cfde29266059fc6c5ed9351d57ac4b44daa0d, 590c9dbe79197d5d9894f92b30446bfb427948de
  - DataFormats/Math/test/Angle0to2piTest.cc:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 69a8c3cfd6f7f36591992b07de90e028ee6972a8, 71953ee36ac3f5caeed01e74da398805f081604c, 2fe825d5791331dbeebc90b73f8cc35ec8812d85, 52c4f4402ab72d079af3b46c3edbd50a17127e8e
    - Deleted: 590c9dbe79197d5d9894f92b30446bfb427948de

cmsbuild commented:
Pull request #26227 was updated. @perrotta, @civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d1ca33ceeda4f0a2a61f4877274cbdb2a77d797b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26227/34433/summary.html
Additional comment: Compilation Warnings: Yes

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26227/34433/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 5710 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3211964
 - DQMHistoTests: Total failures: 7776
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3203984
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

perrotta commented:
I have compared the output of wf 21234.0 of step3, running with and without this PR on the very same step2 root output: results are identical, and in both cases identical to the baseline results of the automatically run jenkins tests.
This confirms that differences arise at step 2 (looking at the jenkins tests outputs, also step1 results are identical when using this PR or without it). According to the tests done by Carl, those differences aren't due to the modification provided by this PR, and therefore a different random sequence in the simulation step remains as the most probable origin.

cvuosalo commented:
+1

kpedro88 commented:
+upgrade

perrotta commented:
+1
- "Phi" and "deltaPhi" utility classes are updated as intended
- Implementation in RECO is consistent with the changes
- Jenkins tests pass, and show differences only in two Phase2 workflows (20034 and 21234); those differences arise at step2 (while reco step reproduces identically when starting from the very same step2 output), and they are compatible with being generated by a different random sequence in the SIM step
   - (Maybe SIM wants to investigate more on it, but for RECO this PR is ready to go)

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d1ca33ceeda4f0a2a61f4877274cbdb2a77d797b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26227/34501/summary.html
Additional comment: Compilation Warnings: Yes

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26227/34501/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 5705 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3211964
 - DQMHistoTests: Total failures: 7776
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3203984
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

ianna commented:
hold

cmsbuild commented:
Pull request has been put on hold by  @ianna
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

ianna commented:
@cvuosalo - gcc 9 does not like it:
```
Warning: Forward declarations of templates with enums as template parameters. The responsible class is: NonTypeTemplateParmDecl 0x1a94af8 </build/yana/Carl/CMSSW_10_6_DEVEL_X_2019-05-03-2300/src/DataFormats/GeometryVector/interface/Phi.h:49:23, col:50> col:32 referenced 'enum Geom::PhiRange' depth 0 index 1 range
`-TemplateArgument expr
  `-DeclRefExpr 0x1a94ab0 <col:40, col:50> 'enum Geom::PhiRange' EnumConstant 0x1a924c0 'MinusPiToPi' 'enum Geom::PhiRange'

Warning: Problems with arguments for forward declaration of class Phi
Warning: Forward declarations of templates with enums as template parameters. The responsible class is: NonTypeTemplateParmDecl 0x1a94af8 </build/yana/Carl/CMSSW_10_6_DEVEL_X_2019-05-03-2300/src/DataFormats/GeometryVector/interface/Phi.h:49:23, col:50> col:32 referenced 'enum Geom::PhiRange' depth 0 index 1 range
`-TemplateArgument expr
  `-DeclRefExpr 0x1a94ab0 <col:40, col:50> 'enum Geom::PhiRange' EnumConstant 0x1a924c0 'MinusPiToPi' 'enum Geom::PhiRange'

Warning: Problems with arguments for forward declaration of class Phi

```

cvuosalo commented:
@ianna: I addressed this issue two weeks ago in this thread:
https://github.com/cms-sw/cmssw/pull/26227#issuecomment-485482466
I added this note to the PR description since it is hard to find in the long thread.

It is not a gcc 9 warning; it is a clang/ROOT warning. As far as I can tell, it is not a real problem.
Could you please unhold this PR?

ianna commented:
unhold

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
@cvuosalo @ianna sorry, this is not a clang issue, it is a compiler warning for what I can see, and should be addressed, no point in merging the code and then opening an issue to fix it.

ianna commented:
@fabiocos - yes, I agree with you.

cvuosalo commented:
@fabiocos: Here is what Philiippe Canal has said about the compiler warning:

On 4/12/19 3:15 PM, Philippe G Canal wrote:
Hi,

The forward declaration are generated by genreflex and stored in a rootmap file.  They are part of the auto-loading and auto-parsing infrastructure.

There is some challenge in getting those (necessary) forward declaration right.  It is possible that despite the warning it might work ....
Cheers,
Philippe.

On 4/17/19 10:16 AM, Philippe G Canal wrote:

Hi Carl,

There is two things that could go wrong.  AutoLoading of library might not be triggered based on the usage of the template instance name.

More importantly, loading in ROOT session (i.e. at the prompt of using gROOT->ProcessLine)  the header file that implement/declares this template instance may not working.

To test the later, start a ROOT prompt, create an instance of or use any class that has a dictionary in that library then finally create/use the template instance with the enum.

Cheers,
Philippe.

********************************************************

I tested in ROOT as Philippe suggested and was able to instantiate the Phi class that uses the enum as its second template parameter. Even a form like `Phi<float>`, where the second template parameter is implicitly given its default value works fine in ROOT. It looks to me like the class works OK, so the warning might be misleading.

I have asked Philippe how we can silence the warning, and I'll post his answer when I get it.

cvuosalo commented:
I also created an alternate implementation that is less elegant but avoids the warning. Converting the non-type template parameter to a typename template parameter silences the warning, though the functionality is identical. I may push this alternate version tomorrow if I don't get another solution from Philippe.

fabiocos commented:
@cvuosalo thank you for following up with this issue.

cvuosalo committed:
Change implementation of Phi 2nd template parameter to avoid spurious genreflex warning

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26227/9691

- This PR adds an extra 220KB to repository

- Found files with invalid states:
  - DataFormats/Math/interface/Angle0to2pi.h:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae, d38e1fc565d8ba3e6ed5240b976f036d6391ce40
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 2fe825d5791331dbeebc90b73f8cc35ec8812d85
    - Deleted: ab1cfde29266059fc6c5ed9351d57ac4b44daa0d, 590c9dbe79197d5d9894f92b30446bfb427948de
  - DataFormats/Math/test/Angle0to2piTest.cc:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 69a8c3cfd6f7f36591992b07de90e028ee6972a8, 71953ee36ac3f5caeed01e74da398805f081604c, 2fe825d5791331dbeebc90b73f8cc35ec8812d85, 52c4f4402ab72d079af3b46c3edbd50a17127e8e
    - Deleted: 590c9dbe79197d5d9894f92b30446bfb427948de

cmsbuild commented:
Pull request #26227 was updated. @perrotta, @civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

cvuosalo commented:
I changed the implementation of the second template parameter of the `Phi` class in `DataFormats/GeometryVector/interface/Phi.h` to avoid the probably spurious "forward declaration" warning from genreflex that is discussed at length above. The new implementation is logically equivalent to the previous one, but, instead of the more appropriate `enum class` constants, it uses dummy structs for the second template parameter to Phi. The advantage is that the apparent sensitivity of genreflex to enums as non-type template parameters is avoided.

If I ever learn of a way to silence the spurious "forward declaration" warning, I might change Phi back to the more elegant `enum` implementation in a future PR.

This latest revision only involves changing the second template parameter of `Phi` from non-type paramter `enum class PhiRange` to a more typical typed template parameter.

cvuosalo commented:
@cmsbuild: please test

cvuosalo commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 753317574bedeba390cb91a1e5988673ad9975a6
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e3a6a9/101/summary.html
Additional comment: 'Compilation

cmsbuild commented:
Comparison job queued.

perrotta commented on DataFormats/GeometryVector/src/classes_def.xml:
```diff
@@ -14,11 +14,11 @@
   </class>
-  <class name="Geom::Phi<double, Geom::PhiRange::MinusPiToPi>" ClassVersion="10">
-   <version ClassVersion="10" checksum="2898804398"/>
+  <class name="Geom::Phi<double, Geom::MinusPiToPi>" ClassVersion="11">
```
There is no reason to increment the class version: please stay with "10"

cvuosalo replied:
Before I updated this file, the compiler gave an error and specified a new checksum and version number. I was only following the instructions from the compiler message. I can try decrementing the version number and see if the compiler complains.

Dr15Jones replied:
You have technically cretated a new class because the template type is different. In that case you should use ClassVersion=3 (we used 10 originally because we needed to be safe reading ROOT files generated before we started adding ClassVersions to the classes_def.xml)

cvuosalo replied:
@Dr15Jones: With the specification of ClassVersion=3, will the code still be able to correctly read old persisted data with the original version of the `Phi` class?

cvuosalo replied:
The answer to my own question is Yes, old persisted data can be read by the new code.

perrotta commented on DataFormats/GeometryVector/src/classes_def.xml:
```diff
@@ -14,11 +14,11 @@
   </class>
-  <class name="Geom::Phi<float, Geom::PhiRange::MinusPiToPi>" ClassVersion="10">
-   <version ClassVersion="10" checksum="4217567924"/>
+  <class name="Geom::Phi<float, Geom::MinusPiToPi>" ClassVersion="11">
```
There is no reason to increment the class version: please stay with "10"

Dr15Jones replied:
The code can't read old files anyway since it is a new C++ type. It just like changing `std::vector<int>` to an `std::vector<MyType>`, ROOT has no ability to automatically deal with the change. 

An explicit IO rule would have to be added to handle conversion for the old type to the new type.

cvuosalo replied:
I did a test and was able to read the old version of `Phi` with the code in this PR, even without an ioread rule. But I think this question is academic because I can't find any place in CMSSW where `Phi` variables are persisted.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e3a6a9/101/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 5711 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3211964
 - DQMHistoTests: Total failures: 7776
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3203984
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

slava77 commented:
> Is the old Geom::Phi persisted anywhere?
> If it is, there should be an effort made to read the old data in a compatible way.

makortel commented on DataFormats/GeometryVector/interface/Phi.h:
```diff
@@ -21,13 +21,14 @@ namespace Geom {
 
-  enum class PhiRange {MinusPiToPi, ZeroTo2pi};
+  struct MinusPiToPi {};  // Dummy struct to indicate -pi to pi range
+  struct ZeroTo2pi{};  // Dummy struct to indicate 0 to 2pi range
```
You could enclose these into another struct if you'd like to keep `PhiRange::MinusPiToPi` etc
```cpp
struct PhiRange {
  struct MinusPiToPi {};  // Dummy struct to indicate -pi to pi range
  struct ZeroTo2pi{};  // Dummy struct to indicate 0 to 2pi range
};
```

cvuosalo replied:
Thanks for the suggestion, but I don't really want to create another dummy `struct`. I'll just stick with these two.
I just learned from Philippe Canal that genreflex cannot correctly handle an `enum class` as a non-type template parameter, so my previous approach could not work.

cvuosalo commented:
@slava77: I have checked all uses of the `Phi` class in CMSSW. There turn out to be not many. I could not find any case where a `Phi` variable is written or stored anywhere.

Just in case, I will do some tests of reading and writing the different versions of the `Phi` class.

slava77 commented:
> @slava77: I have checked all uses of the `Phi` class in CMSSW. There turn out to be not many. I could not find any case where a `Phi` variable is written or stored anywhere.

OK. Perhaps it's reasonably safe to assume it wasn't saved ever in anything useful and postpone the possible need to recover to a later time if a problem is reported.

cvuosalo committed:
Change ClassVersion to 3 since revised Phi template is effectively a new class

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26227/9783

- This PR adds an extra 92KB to repository

- Found files with invalid states:
  - DataFormats/Math/interface/Angle0to2pi.h:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae, d38e1fc565d8ba3e6ed5240b976f036d6391ce40
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 2fe825d5791331dbeebc90b73f8cc35ec8812d85
    - Deleted: ab1cfde29266059fc6c5ed9351d57ac4b44daa0d, 590c9dbe79197d5d9894f92b30446bfb427948de
  - DataFormats/Math/test/Angle0to2piTest.cc:
    - Added: 5dea68285f63c491bef353b01cee55757930a4ae
    - Modified: 533a9ee4efe7cf8757445f7e0e4ee36457743405, 69a8c3cfd6f7f36591992b07de90e028ee6972a8, 71953ee36ac3f5caeed01e74da398805f081604c, 2fe825d5791331dbeebc90b73f8cc35ec8812d85, 52c4f4402ab72d079af3b46c3edbd50a17127e8e
    - Deleted: 590c9dbe79197d5d9894f92b30446bfb427948de

cmsbuild commented:
Pull request #26227 was updated. @perrotta, @civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

cvuosalo commented:
@slava77: Concerning backward compatibility, I wrote the old version of `Phi` to a ROOT file and then read it out successfully with the code in this PR, so in the unlikely case there is some existing ROOT file containing the old `Phi` class, it should be readable with the new code.

cvuosalo commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 3265dea9b51e768e3232275fe0ef047139db41b6
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e3a6a9/218/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e3a6a9/218/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 5766 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3212004
 - DQMHistoTests: Total failures: 7765
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3203905
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

cvuosalo commented:
As before, the comparison test differences are almost entirely in two Phase 2 simulation workflows which have different random sequences in the SIM step. These differences are not due to this PR.

I think all issues in this PR have been addressed, so I hope it can be soon approved.

cvuosalo commented:
+1

perrotta commented:
+1
- Changes since last reco signature in https://github.com/cms-sw/cmssw/pull/26227#issuecomment-488596718 are consequential to the review comments arrived later on
- Worries about possible backward incompatibilities seem can be dissipated by the checks in https://github.com/cms-sw/cmssw/pull/26227#issuecomment-492430015

fabiocos commented:
@civanch @kpedro88 do you have further comments?

civanch commented:
+1

kpedro88 commented:
+upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

kpedro88 commented:
+1

slava77 commented on DataFormats/Math/interface/deltaPhi.h:
```diff
@@ -59,4 +61,76 @@ struct DeltaPhi {
+  // calling fmod to improve performance.
+
+  template <class valType>
+  inline constexpr valType make0To2pi(valType angle) {
```
commenting here instead of in #27637 

about angle0to2pi
IMHO, it would be more consistent to move it up in this file to the reco namespace and rename to ```reduceRange0To2pi```
Also, if it's implementation is really faster, replace the current ```reduceRange```  of [-pi,pi] with the faster variant.

ianna replied:
@slava77 and @cvuosalo - can we think of other then reco namespace? It would be confusing if the DD code uses reco::

cvuosalo replied:
Concerning the possible replacement of `reduceRange`: Since `make0To2pi` checks for floating-point value drift over repeated calculations and `reduceRange` does not, there is a good possibility that comparison tests for replacing `reduceRange` would show small differences. `make0To2pi` should provide more correct results, but it would be required to show that the comparison test differences were correct, appropriate, and desirable, which might be a challenge. I did not replace `reduceRange` in this PR for this very reason.

Perhaps replacement could be done on a step-by-step basis by developers familiar with the applications getting changed so that the comparison differences would be easier to understand.

cvuosalo replied:
I'd suggest putting everything in the file in the `delta_phi` namespace to match the header file name, but I don't have a strong opinion about namespace naming.

slava77 replied:
About the small differences, it should be acceptable, considering the expected speedup and improved numerical precision (unless perhaps a PR comes after the deadline with an "urgent" tag and some unrelated changes inside).
